### ISSUE TITLE
compute: Avoid overflow when reducing plans

### ISSUE THREAD

**Issue by jkosh44** [MaterializeInc/materialize#14695]
### What version of Materialize are you using?

`main`

### How did you install Materialize?

Built from source

### What is the issue?

https://github.com/MaterializeInc/materialize/blob/main/src/compute/src/render/reduce.rs uses a lot of unchecked addition, for example
https://github.com/MaterializeInc/materialize/blob/47483b2c089f1e9470991b2ef271ff3465e88ab4/src/compute/src/render/reduce.rs#L813-L817
https://github.com/MaterializeInc/materialize/blob/47483b2c089f1e9470991b2ef271ff3465e88ab4/src/compute/src/render/reduce.rs#L794-L795
These can all result in overflow panics.

Also it uses an `as` cast which can result in incorrect results due to truncation
https://github.com/MaterializeInc/materialize/blob/47483b2c089f1e9470991b2ef271ff3465e88ab4/src/compute/src/render/reduce.rs#L1272-L1275

Any easy way to create a panic from this is from the following query:
```
materialize=> CREATE TABLE t (a int8);
CREATE TABLE
materialize=> INSERT INTO t VALUES (13), (32767), (2147483647), (9223372036854775807);
INSERT 0 4
materialize=> SELECT STDDEV(a) FROM t;
```

### Relevant log output

```source
thread 'timely:work-0' panicked at 'attempt to add with overflow', /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/arith.rs:786:1
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:48:5
   3: <i128 as core::ops::arith::AddAssign>::add_assign
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/arith.rs:779:51
   4: <i128 as core::ops::arith::AddAssign<&i128>>::add_assign
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/internal_macros.rs:127:17
   5: <mz_compute::render::reduce::AccumInner as differential_dataflow::difference::Semigroup>::plus_equals
             at ./src/compute/src/render/reduce.rs:813:17
   6: <mz_compute::render::reduce::Accum as differential_dataflow::difference::Semigroup>::plus_equals
             at ./src/compute/src/render/reduce.rs:955:9
   7: differential_dataflow::difference::vector::<impl differential_dataflow::difference::Semigroup for alloc::vec::Vec<R>>::plus_equals
             at /home/joe/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/78e7319/src/difference.rs:242:17
   8: differential_dataflow::consolidation::consolidate_updates_slice
             at /home/joe/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/78e7319/src/consolidation.rs:125:17
   9: differential_dataflow::consolidation::consolidate_updates_from
             at /home/joe/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/78e7319/src/consolidation.rs:95:18
  10: differential_dataflow::consolidation::consolidate_updates
             at /home/joe/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/78e7319/src/consolidation.rs:86:5
  11: <differential_dataflow::collection::Collection<G,D1,R> as mz_timely_util::operator::CollectionExt<G,D1,R>>::explode_one::{{closure}}::{{closure}}::{{closure}}
             at ./src/timely-util/src/operator.rs:295:25
  12: timely::dataflow::operators::generic::handles::InputHandleCore<T,D,P>::for_each
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/dataflow/operators/generic/handles.rs:97:13
  13: <differential_dataflow::collection::Collection<G,D1,R> as mz_timely_util::operator::CollectionExt<G,D1,R>>::explode_one::{{closure}}::{{closure}}
             at ./src/timely-util/src/operator.rs:289:21
  14: <timely::dataflow::stream::StreamCore<G,D1> as timely::dataflow::operators::generic::operator::Operator<G,D1>>::unary::{{closure}}::{{closure}}
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/dataflow/operators/generic/operator.rs:401:17
  15: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build::{{closure}}::{{closure}}
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/dataflow/operators/generic/builder_rc.rs:123:31
  16: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/dataflow/operators/generic/builder_rc.rs:162:26
  17: <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/dataflow/operators/generic/builder_raw.rs:204:9
  18: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:684:30
  19: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:344:26
  20: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:310:17
  21: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:684:30
  22: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:344:26
  23: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:310:17
  24: timely::worker::Wrapper::step::{{closure}}
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/worker.rs:754:57
  25: core::option::Option<T>::map
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/option.rs:929:29
  26: timely::worker::Wrapper::step
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/worker.rs:754:26
  27: timely::worker::Worker<A>::step_or_park
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/worker.rs:396:38
  28: mz_compute::server::Worker<A>::run_client
             at ./src/compute/src/server.rs:385:13
  29: mz_compute::server::Worker<A>::setup_channel_and_run_client
             at ./src/compute/src/server.rs:359:9
  30: mz_compute::server::Worker<A>::run
             at ./src/compute/src/server.rs:224:21
  31: mz_compute::server::serve::{{closure}}
             at ./src/compute/src/server.rs:117:13
  32: timely::execute::execute_from::{{closure}}
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/execute.rs:388:22
  33: timely_communication::initialize::initialize_from::{{closure}}
             at /home/joe/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/communication/src/initialize.rs:316:33
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by jkosh44** [comment]
The silent data loss due to truncation is REALLY hard to reproduce because it requires A LOT of data. I tried to write the following python script to reproduce the error, but at the rate it was running it seemed like it might take a couple days to complete.

```python
from ast import Str
import subprocess


def main():
    execute_sql("DROP TABLE IF EXISTS t")
    execute_sql("CREATE TABLE t (a int4)");
    for _ in range(4294967299):
        execute_sql("INSERT INTO t VALUES (2147483647)", True)
    execute_sql("SELECT SUM(a) FROM t")
    


def execute_sql(query: Str, silent: bool = False):
    command = "psql -U materialize -h localhost -p 6875 materialize -c".split()
    command.append(query)
    process = subprocess.Popen(command, stdout=subprocess.PIPE)
    output, error = process.communicate()
    if not silent:
       print(output.decode("utf-8"))

if __name__ == "__main__":
    main()
```

**Comment by antiguru** [comment]
Thanks for the report, this is super helpful. I'll make sure to prioritize it.

**Comment by frankmcsherry** [comment]
Fwiw, some of these are meant to panic. We have to make the call between "incorrect" and "panics", and can try and figure out what we'd like most. DD should work fine if these are wrapping operations, except for the diff addition (which when it wraps will present as a weirdly absent record). Many of these errors were judged to be tolerable when we were defining the accumulation behavior, and MZ's non-interest in arbitrary accuracy.

tl;dr: we should figure out what the intended behavior is here, as panicking may be better than wrong results.

**Comment by jkosh44** [comment]
Unfortunately though overflowing does not necessarily mean there isn't a correct answer. For example the following overflows and panics, even though it should be easily representable.
```
materialize=> CREATE TABLE t (a int8);
CREATE TABLE
materialize=> INSERT INTO t VALUES (13), (32767), (2147483647), (9223372036854775807);
INSERT 0 4
materialize=> DELETE FROM t;
DELETE 4
materialize=> INSERT INTO t VALUES (1), (2);
INSERT 0 2
materialize=> SELECT STDDEV(a) FROM t;
```

I'm not super familiar with this code, but my guess is that the insertions are added and the deletions are subtracted, so we overflow before the deletions happen. If so we probably want to consider using arbitrary precision types for the accumulation and only check for overflow at the very end.

**Comment by lluki** [comment]
> Fwiw, some of these are meant to panic. We have to make the call between "incorrect" and "panics", and can try and figure out what we'd like most. DD should work fine if these are wrapping operations, except for the diff addition (which when it wraps will present as a weirdly absent record). 

The diff addition only happens in DD consolidate, which in turn uses the default `+=` operator for signed integers. (or did i overlook a case?). So I think we panic already on overflowing diffs. Although I'm having a hard time coming up with a SQL query that would generate such large diffs without materializing the input (at which point we'd probably run out of memory first). 

The optimizer (correctly) cuts off if it gets too large `EXPLAIN SELECT 1 from generate_series(1,100000);` doesn't create a constant.

> Many of these errors were judged to be tolerable when we were defining the accumulation behavior, and MZ's non-interest in arbitrary accuracy.
> 
> tl;dr: we should figure out what the intended behavior is here, as panicking may be better than wrong results.

I think panicking is better than wrong results, and it is what most operations do in reduce anyway. Except of course the downcast that joe has found. Unless we go for a bigint accumulator.

**Comment by frankmcsherry** [comment]
> So I think we panic already on overflowing diffs.

We should wrap on overflowing diffs (the behavior in release mode). It's somewhat important that we do this, as otherwise we don't have a group (at least, we shouldn't do saturating math). We could change it to panic, at some runtime cost, but yes it should only be visible for computations over what would be ~2^63 records in SQL.

You can probably make MZ encounter this by doing a bunch of self-joins with a collection containing multiple copies of the same record; enough that would cause unary SQL systems to take forever, but which DD does efficiently enough to risk wrapping.

**Comment by lluki** [comment]
Oh, release mode turns off the overflow checking TIL. Ok, now everything makes sense. So then my suggestion is to add the explicitly checked versions in `reduce.rs`.

I'm unsure about the diff handling in DD, maybe we should just add the checked version there too, I don't think checking for overflow is very expensive.

**Comment by lluki** [comment]
Let's ignore a second the potential diff overflow that @frankmcsherry pointed out. The initial overflow reported by @jkosh44 happens because we are using a "i128" for the accumulation of float values:
https://github.com/MaterializeInc/materialize/blob/47483b2c089f1e9470991b2ef271ff3465e88ab4/src/compute/src/render/reduce.rs#L714-L716

That seems like not too great of an idea, because it will inevitable round away small and big values:

```sql
# Quick dance to avoid optimizations
materialize=> CREATE TABLE t (a int8);
CREATE TABLE
materialize=> INSERT INTO t VALUES (1);
INSERT 0 1

# This query will not be optimized away and go through the reduce operator. Wrong result.
materialize=> SELECT SUM(1e-39::double) FROM t;
 sum 
-----
   0
(1 row)

# This query will be optimized away and produces the correct result
materialize=> SELECT SUM(1e-39::double);
  sum  
-------
 1e-39
(1 row)

# Same again for really big values. Wrong result
materialize=> SELECT SUM(1e38::double) FROM t;
          sum           
------------------------
 1.0141204801825835e+31
(1 row)

# optimized away, correct.
materialize=> SELECT SUM(1e38::double);
  sum  
-------
 1e+38
(1 row)

materialize=>
```

I dont think there is a good solution to this, except going to do the aggregation in double precision domain. But that conflicts with the associativity requirement. So, arbitrary precision?

**Comment by uce** [comment]
(I removed the triage label as this is in progress)

**Comment by lluki** [comment]
There might be a smart solution for deterministic (associative + commutative) floating point aggregation: https://arxiv.org/pdf/1802.09883.pdf  I still need to read the paper, but it sounds promising!

**Comment by lluki** [comment]
Here is some more probably for user unexpected behavior

```
# PSQL creates an error on overflow
lukas=> CREATE TABLE d (a double precision);
CREATE TABLE
lukas=> insert into d values (1e308),(1e308);
INSERT 0 2
lukas=> select sum(a) from d;
ERROR:  value out of range: overflow
```

```
# Materialize in release mode, we do wrapping arithmetic
materialize=> select * from d;
   a   
-------
 1e+31
 1e+31
(2 rows)

materialize=> select SUM(a) from d;
           sum           
-------------------------
 -2.8240960365167115e+29
(1 row)

# normal addition is not affected
materialize=> CREATE TABLE d1 (a double, b double);
CREATE TABLE
materialize=> insert into d1 values (1e31, 1e31);
INSERT 0 1
materialize=> SELECT a + b from d1;
 ?column? 
----------
    2e+31
(1 row)


# We accept doubles that are out of range for our aggregator
materialize=> DELETE FROM d;
DELETE 2
materialize=> insert into d values (1e200);
ERROR:  "1E200" is out of range for type numeric: exceeds maximum precision 39
materialize=> insert into d values (1e39);
ERROR:  "1E39" is out of range for type numeric: exceeds maximum precision 39
materialize=> insert into d values (1e38);
INSERT 0 1
materialize=> SELECT * FROM d;
   a   
-------
 1e+38
(1 row)

materialize=> SELECT SUM(a) FROM d;
          sum           
------------------------
 1.0141204801825835e+31
(1 row)
```

**Comment by lluki** [comment]
I'm about to close this issue with the following outcomes: 
 * PR #15187  adds explicit wrapping arithmetic such that envd doesnt crash in debug mode.
 * Issue #15186  tracks the underlying issue: That our floating point aggregator is not precise enough.

**Comment by frankmcsherry** [comment]
This works for me. I think it is still unknown whether we want to present overflowed results or crash instead. We had at least one example of postgres presenting overflowed results, so it isn't religion that we cannot present them. But we might eventually conclude that "no wrong answers" is more important than "never crashes".

**Comment by antiguru** [comment]
Closed by #15187. Follow-up worked tracked seperately.
